### PR TITLE
[WIP] Document field init shorthand

### PR DESCRIPTION
**Pull request by ghost**
- [x] Update grammar (commit: [#422895a](https://github.com/hngiang/rust/commit/422895ac8b7486dec2f659ab97c1036f89f6e173))
- [x] Add section [the book](https://github.com/rust-lang/rust/blob/master/src/doc/book/structs.md) (commit: [#66bc3ca](https://github.com/hngiang/rust/commit/66bc3ca0adc13cc5b33270effbc9b798d5e48114)) and [the new book](https://github.com/rust-lang/book) (PR: [#405](https://github.com/rust-lang/book/pull/405))
- [x] Add [reference](https://github.com/rust-lang/rust/blob/master/src/doc/reference.md) explain (commit: [#00650de](https://github.com/rust-lang/rust/commit/00650de1a8c323826eae558c304838b2795a3648))
- [x] Add examples to Rust By Example (PR: [#837](https://github.com/rust-lang/rust-by-example/pull/837))

- [ ] Uncomplete a changelog entry

### CODE REVIEW

File: src/doc/book/structs.md [hunk 1]
```diff
@@ -117,6 +117,28 @@ fn main() {
 }
 ```
 
+We can initializing a data structure (struct, enum, union) with named fields, by writing `fieldname` as a shorthand for `fieldname: fieldname`. This allows a compact syntax for initialization, with less duplication:
```
Comment (GuillaumeGomez) [inline src/doc/book/structs.md]: "intializing" -> initialize

File: src/doc/reference.md [hunk 2]
```diff
@@ -2757,6 +2757,19 @@ let base = Point3d {x: 1, y: 2, z: 3};
 Point3d {y: 0, z: 10, .. base};
 ```
 
+#### Struct field init shorthand
+
+When initializing a data structure (struct, enum, union) with named fields, allow writing `fieldname` as a shorthand for `fieldname: fieldname`. This allows a compact syntax for initialization, with less duplication.
+
+In the initializer for a `struct` with named fields, a `union` with named fields, or an enum variant with named fields, accept an identifier `field` as a shorthand for `field: field`.
```
Comment (GuillaumeGomez) [inline src/doc/reference.md]: "a union" -> "an union"

Comment (Havvy) [inline src/doc/reference.md]: As an English speaker, "a union" sounds write to me. Silly little quirks in this language...

File: src/doc/book/structs.md (hunk 1 above)
Comment (Havvy) [inline src/doc/book/structs.md]: This sentence doesn't read write to me at all. I would replace all of `a data structure (struct, enum, union) with named fields` with `named fields`.

I would also add a sentence saying that a variable of the name of the field also has to be in scope.

File: src/doc/reference.md (hunk 2 above)
Comment (KalitaAlexey) [inline src/doc/reference.md]: @GuillaumeGomez,
It is actually "a union".
Proof: http://www.businessdictionary.com/definition/union.html

File: src/doc/book/structs.md (hunk 1 above)
Comment (KalitaAlexey) [inline src/doc/book/structs.md]: It seems wrong to me.
I would write:
Initialization of a data structure can be simplified if fields of the data structure are initialized with variables which has same names as the fields.

File: src/doc/reference.md (hunk 2 above)
Comment (GuillaumeGomez) [inline src/doc/reference.md]: English is way too complicated...

Comment (KalitaAlexey) [inline src/doc/reference.md]: @GuillaumeGomez,
Please read http://english.stackexchange.com/questions/266309/why-is-union-an-exception-to-the-a-an-rule
It is actually simple rule.

Comment (GuillaumeGomez) [inline src/doc/reference.md]: It's always a simple rule, but there are a lot of them, which makes the whole complicated.

File: src/doc/book/structs.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/book/structs.md]: I agree with @KalitaAlexey 's suggestion.

**Comment by steveklabnik** [comment]
Thank you so much @hngiang ! Could you also wrap this to 80 columns, please?

**Comment by hngiang** [comment]
@steveklabnik: I haven't been available these days due to the big holiday in my country, so I couldn't reach the topic. At the moment, I have some troubles in work and cannot fix these problems, I will ask @phungleson for help. Thanks.

**Comment by steveklabnik** [comment]
@hngiang no worries! I totally understand; I hope your holiday went well! Good luck with the work situation. I will close in favor of https://github.com/rust-lang/rust/pull/39459

**Pull request closed without merging**